lib/ecmascript.ts

```diff
@@ -2861,20 +2861,50 @@ export function GetIANATimeZonePreviousTransition(epochNanoseconds: JSBI, id: st
   return result;
 }
 
+// ts-prune-ignore-next TODO: remove this after tests are converted to TS
+export function parseFromEnUsFormat(datetime: string) {
+  const parts = datetime.split(/[^\w]+/);
+
+  if (parts.length !== 7) {
+    throw new RangeError(`expected 7 parts in "${datetime}`);
+  }
+
+  const month = +parts[0];
+  const day = +parts[1];
+  let year = +parts[2];
+  const era = parts[3].toUpperCase();
+  if (era === 'B' || era === 'BC') {
+    year = -year + 1;
+  } else if (era !== 'A' && era !== 'AD') {
+    throw new RangeError(`Unknown era ${era} in "${datetime}`);
+  }
+  let hour = +parts[4];
+  if (hour === 24) {
+    hour = 0;
+  }
+  const minute = +parts[5];
+  const second = +parts[6];
+
+  if (
+    !NumberIsFinite(year) ||
+    !NumberIsFinite(month) ||
+    !NumberIsFinite(day) ||
+    !NumberIsFinite(hour) ||
+    !NumberIsFinite(minute) ||
+    !NumberIsFinite(second)
+  ) {
+    throw new RangeError(`Invalid number in "${datetime}`);
+  }
+
+  return { year, month, day, hour, minute, second };
+}
+
 // ts-prune-ignore-next TODO: remove this after tests are converted to TS
 export function GetFormatterParts(timeZone: string, epochMilliseconds: number) {
   const formatter = getIntlDateTimeFormatEnUsForTimeZone(timeZone);
   // Using `format` instead of `formatToParts` for compatibility with older clients
   const datetime = formatter.format(new Date(epochMilliseconds));
-  const [month, day, year, era, hour, minute, second] = datetime.split(/[^\w]+/);
-  return {
-    year: era.toUpperCase().startsWith('B') ? -year + 1 : +year,
-    month: +month,
-    day: +day,
-    hour: hour === '24' ? 0 : +hour, // bugs.chromium.org/p/chromium/issues/detail?id=1045791
```

Comment:
It'd be nice to preserve this link in a comment in the new code, so that we know why the check for 24 is present.

Reply:
Thanks for the review, I added that comment back.